optuna/_gp/batched_lbfgsb.py

```diff
@@ -70,30 +72,39 @@ def _func_and_grad(x: np.ndarray) -> tuple[float, np.ndarray]:
         xs_opt[i] = x_opt
         fvals_opt[i] = fval_opt
         n_iterations[i] = info["nit"]
-        is_remaining_batch[i] = False
 
     greenlets = [greenlet(run) for _ in range(batch_size)]
-    x_batched = [gl.switch(i) for i, gl in enumerate(greenlets)]
-
-    while np.any(is_remaining_batch):
-        remaining_batch_indices = np.where(is_remaining_batch)[0]
-        fvals, grads = func_and_grad(np.asarray(x_batched), np.asarray(remaining_batch_indices))
-
-        x_batched = []
-        next_greenlets = []
-        for i, gl in enumerate(greenlets):
-            x = gl.switch((fvals[i], grads[i]))
-            if x is None:
-                continue
-            x_batched.append(x)
-            next_greenlets.append(gl)
-        greenlets = next_greenlets
+    x_and_argsT_pairs = [gl.switch(i) for i, gl in enumerate(greenlets)]
+
+    while x_and_argsT_pairs:
+        x_batched_list = [pair[0] for pair in x_and_argsT_pairs if pair is not None]
+        args_tuple_transposed = [pair[1] for pair in x_and_argsT_pairs if pair is not None]
+        x_batched = np.array(x_batched_list)
+
+        if args_tuple_transposed and all(a is not None for a in args_tuple_transposed):
+            current_args_tuple = tuple(zip(*args_tuple_transposed))
+            fvals, grads = func_and_grad(x_batched, *current_args_tuple)
+        else:
+            fvals, grads = func_and_grad(x_batched)
+
+        results = [gl.switch((fvals[i], grads[i])) for i, gl in enumerate(greenlets)]
+
+        live_pairs_and_greenlets = [
+            (pair, gl) for pair, gl in zip(results, greenlets) if pair is not None
+        ]
+
+        if not live_pairs_and_greenlets:
+            break
+
+        x_and_argsT_pairs, greenlets = map(list, zip(*live_pairs_and_greenlets))
+
     return xs_opt, fvals_opt, n_iterations
 
 
 def batched_lbfgsb(
-    func_and_grad: Callable[[np.ndarray, np.ndarray], tuple[np.ndarray, np.ndarray]],
+    func_and_grad: Callable[..., tuple[np.ndarray, np.ndarray]],
     x0_batched: np.ndarray,
+    args_tuple: tuple[Any] | None = None,
```

Comment:
How about use an empty tuple as a default value?

```suggestion
    args_tuple: tuple[Any] = (),
```

Reply:
Also looks better for me, so I fixed it. Thank you for your feedback!